test/integration/udproute_test.go

```diff
@@ -0,0 +1,361 @@
+	require.NoError(t, err)
+
+	t.Log("verifying that the Gateway gets linked to the route via status")
+	udpeventuallyGatewayIsLinkedInStatus(t, c, ns.Name, udproute.Name)
+
+	t.Log("verifying that creating the GatewayClass again triggers reconciliation of UDPRoutes and the route becomes available again")
+	t.Logf("checking DNS to resolve via UDPIngress %s", udproute.Name)
+	assert.Eventually(t, func() bool {
+		_, err := resolver.LookupHost(ctx, "kernel.org")
+		return err == nil
+	}, ingressWait, waitTick)
+
+	t.Log("deleting the Gateway")
+	oldGWName := gw.Name
+	require.NoError(t, c.GatewayV1alpha2().Gateways(ns.Name).Delete(ctx, gw.Name, metav1.DeleteOptions{}))
+
+	t.Log("verifying that the Gateway gets unlinked from the route via status")
+	udpeventuallyGatewayIsUnlinkedInStatus(t, c, ns.Name, udproute.Name)
+
+	t.Log("verifying that the data-plane configuration from the UDPRoute gets dropped with the Gateway now removed")
+	assert.Eventually(t, func() bool {
+		_, err := resolver.LookupHost(ctx, "kernel.org")
+		return err != nil
+	}, ingressWait, waitTick)
+
+	t.Log("putting the Gateway back")
+	gw = &gatewayv1alpha2.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: oldGWName,
+			Annotations: map[string]string{
+				unmanagedAnnotation: "true", // trigger the unmanaged gateway mode
+			},
+		},
+		Spec: gatewayv1alpha2.GatewaySpec{
+			GatewayClassName: gatewayv1alpha2.ObjectName(gwc.Name),
+			Listeners: []gatewayv1alpha2.Listener{{
+				Name:     "udp",
+				Protocol: gatewayv1alpha2.UDPProtocolType,
+				Port:     gatewayv1alpha2.PortNumber(9999),
+			}},
+		},
+	}
+	gw, err = c.GatewayV1alpha2().Gateways(ns.Name).Create(ctx, gw, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	t.Log("verifying that the Gateway gets linked to the route via status")
+	udpeventuallyGatewayIsLinkedInStatus(t, c, ns.Name, udproute.Name)
+
+	t.Log("verifying that creating the Gateway again triggers reconciliation of UDPRoutes and the route becomes available again")
+	t.Logf("checking DNS to resolve via UDPIngress %s", udproute.Name)
+	assert.Eventually(t, func() bool {
+		_, err := resolver.LookupHost(ctx, "kernel.org")
+		return err == nil
+	}, ingressWait, waitTick)
+
+	t.Log("deleting both GatewayClass and Gateway rapidly")
+	require.NoError(t, c.GatewayV1alpha2().GatewayClasses().Delete(ctx, gwc.Name, metav1.DeleteOptions{}))
+	require.NoError(t, c.GatewayV1alpha2().Gateways(ns.Name).Delete(ctx, gw.Name, metav1.DeleteOptions{}))
+
+	t.Log("verifying that the Gateway gets unlinked from the route via status")
+	udpeventuallyGatewayIsUnlinkedInStatus(t, c, ns.Name, udproute.Name)
+
+	t.Log("verifying that the data-plane configuration from the UDPRoute does not get orphaned with the GatewayClass and Gateway gone")
+	assert.Eventually(t, func() bool {
+		_, err := resolver.LookupHost(ctx, "kernel.org")
+		return err != nil
+	}, ingressWait, waitTick)
+}
+
+// TODO consolidate shared util gateway linked funcs
```

Comment:
Looks like we should get this TODO resolved as a part of this PR.

Reply:
For this and the templating, I was planning on deferring to either TCPRoute or TLSRoute. These are less TODOs than they are DRYREFACTORNEXTREVISIT since I'll be passing over the same code again in short order. TODO makes them easier for me to spot though, since my editor highlights that.